File: backend/expressjs/src/engine/nested-write.ts
import type { Queryable, Store } from "../store/postgres.js";
import { queryRow, queryRows, exec } from "../store/postgres.js";
import type { Entity } from "../metadata/types.js";
import type { Registry } from "../metadata/registry.js";
import { getField, fieldNames } from "../metadata/types.js";
import type { ErrorDetail } from "./errors.js";
import { AppError, validationError } from "./errors.js";
import type { RelationWrite } from "./writer.js";
import {
  buildInsertSQL,
  buildUpdateSQL,
  validateFields,
  separateFieldsAndRelations,
} from "./writer.js";
import { executeChildWrite } from "./diff.js";
import { evaluateRules } from "./rules.js";
import { evaluateStateMachines } from "./state-machine.js";
import { triggerWorkflows } from "./workflow.js";
import { fireAsyncWebhooks, fireSyncWebhooks } from "./webhook.js";
import { getInstrumenter } from "../instrument/instrument.js";

export interface WritePlan {
  isCreate: boolean;
  entity: Entity;
  fields: Record<string, any>;
  id: any;
  childOps: RelationWrite[];
  user: { id: string; roles: string[] } | null;
}

export function planWrite(
  entity: Entity,
  registry: Registry,
  body: Record<string, any>,
  existingID: any | null,
): WritePlan {
  const [fields, relWrites, unknownKeys] = separateFieldsAndRelations(
    entity,
    registry,
    body,
  );

  // Reject unknown keys
  if (unknownKeys.length > 0) {
    const errs: ErrorDetail[] = unknownKeys.map((key) => ({
      field: key,
      rule: "unknown",
      message: `Unknown field or relation: ${key}`,
    }));
    throw validationError(errs);
  }

  const isCreate = existingID == null;

  const validationErrs = validateFields(entity, fields, isCreate);
  if (validationErrs.length > 0) {
    throw validationError(validationErrs);
  }

  const plan: WritePlan = {
    isCreate,
    entity,
    fields,
    id: existingID,
    childOps: Array.from(relWrites.values()),
    user: null,
  };

  return plan;
}

export async function executeWritePlan(
  store: Store,
  registry: Registry,
  plan: WritePlan,
): Promise<Record<string, any>> {
  const span = getInstrumenter().startSpan("engine", "writer", "nested_write.execute");
  span.setEntity(plan.entity.name, plan.id ?? undefined);
  span.setMetadata("operation", plan.isCreate ? "create" : "update");
  const client = await store.beginTx();
  try {
    // Evaluate rules (field → expression → computed)
    let old: Record<string, any> = {};
    if (!plan.isCreate) {
      try {
        old = await fetchRecord(client, plan.entity, plan.id);
      } catch {
        // ignore if old record not found
      }
    }

    const ruleErrs = evaluateRules(
      registry,
      plan.entity.name,
      "before_write",
      plan.fields,
      old,
      plan.isCreate,
    );
    if (ruleErrs.length > 0) {
      throw validationError(ruleErrs);
    }

    // Evaluate state machines (after rules, before SQL write)
    const smErrs = evaluateStateMachines(
      registry,
      plan.entity.name,
      plan.fields,
      old,
      plan.isCreate,
    );
    if (smErrs.length > 0) {
      throw validationError(smErrs);
    }

    // Auto-generate slug if configured
    await autoGenerateSlug(client, plan.entity, plan.fields, plan.isCreate, old, plan.id);

    // Resolve file fields: UUID string → JSONB metadata object
    await resolveFileFields(client, plan.entity, plan.fields);

    let parentID: any;

    if (plan.isCreate) {
      const [sql, params] = buildInsertSQL(plan.entity, plan.fields);
      const row = await queryRow(client, sql, params);
      parentID = row[plan.entity.primary_key.field];
    } else {
      parentID = plan.id;
      const [sql, params] = buildUpdateSQL(plan.entity, plan.id, plan.fields);
      if (sql) {
        await exec(client, sql, params!);
      }
    }

    for (const childOp of plan.childOps) {
      await executeChildWrite(client, registry, parentID, childOp);
    }

    // Pre-commit: fire sync (before_write) webhooks
    const action = plan.isCreate ? "create" : "update";
    await fireSyncWebhooks(client, registry, "before_write", plan.entity.name, action, plan.fields, old, plan.user);

    await client.query("COMMIT");

    // Fetch the full record
    const result = await fetchRecord(store.pool, plan.entity, parentID);

    // Post-commit: trigger workflows for state transitions
    const machines = registry.getStateMachinesForEntity(plan.entity.name);
    for (const sm of machines) {
      const oldState = old[sm.field] != null ? String(old[sm.field]) : "";
      const newState = result[sm.field] != null ? String(result[sm.field]) : "";
      if (oldState !== newState && newState !== "") {
        triggerWorkflows(store, registry, plan.entity.name, sm.field, newState, result, parentID).catch((err) => {
          console.error(`ERROR: triggering workflows for ${plan.entity.name}.${sm.field} -> ${newState}:`, err);
        });
      }
    }

    // Post-commit: fire async (after_write) webhooks
    fireAsyncWebhooks(store, registry, "after_write", plan.entity.name, action, result, old, plan.user);

    span.setStatus("ok");
    return result;
  } catch (err) {
    await client.query("ROLLBACK");
    span.setStatus("error");
    span.setMetadata("error", (err as Error).message);
    throw err;
  } finally {
    client.release();
    span.end();
  }
}

export async function fetchRecord(
  q: Queryable,
  entity: Entity,
  idOrSlug: any,
): Promise<Record<string, any>> {
  const columns = fieldNames(entity);
  if (entity.soft_delete && !getField(entity, "deleted_at")) {
    columns.push("deleted_at");
  }

  const softDeleteClause = entity.soft_delete ? " AND deleted_at IS NULL" : "";

  // If entity has a slug config and the param doesn't look like the PK type, try slug first
  if (entity.slug && typeof idOrSlug === "string" && !looksLikePK(entity, idOrSlug)) {
    const slugSql = `SELECT ${columns.join(", ")} FROM ${entity.table} WHERE ${entity.slug.field} = $1${softDeleteClause}`;
    try {
      return await queryRow(q, slugSql, [idOrSlug]);
    } catch {
      // slug lookup failed, fall through to PK lookup
    }
  }

  // Default: look up by primary key
  const sql = `SELECT ${columns.join(", ")} FROM ${entity.table} WHERE ${entity.primary_key.field} = $1${softDeleteClause}`;
  return queryRow(q, sql, [idOrSlug]);
}

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const INT_RE = /^\d+$/;

function looksLikePK(entity: Entity, value: string): boolean {
  const pkType = entity.primary_key.type;
  if (pkType === "uuid") return UUID_RE.test(value);
  if (pkType === "int" || pkType === "integer" || pkType === "bigint") return INT_RE.test(value);
  return false; // string PKs — can't distinguish, always try slug first
}

export function slugify(text: string): string {
  return text
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "") // strip accents
    .replace(/[^a-z0-9]+/g, "-")     // non-alphanumeric → hyphen
    .replace(/^-+|-+$/g, "")         // trim leading/trailing hyphens
    .replace(/-{2,}/g, "-");          // collapse multiple hyphens
}

export async function generateUniqueSlug(
  q: Queryable,
  entity: Entity,
  baseSlug: string,
  excludeId?: any,
): Promise<string> {
  const slugField = entity.slug!.field;
  const softDeleteClause = entity.soft_delete ? " AND deleted_at IS NULL" : "";
  const excludeClause = excludeId != null ? ` AND ${entity.primary_key.field} != $2` : "";
  const params: any[] = [baseSlug];
  if (excludeId != null) params.push(excludeId);

  const checkSql = `SELECT 1 FROM ${entity.table} WHERE ${slugField} = $1${softDeleteClause}${excludeClause} LIMIT 1`;
  const rows = await queryRows(q, checkSql, params);
  if (rows.length === 0) return baseSlug;

  // Append incrementing suffix
  for (let i = 2; i <= 100; i++) {
    const candidate = `${baseSlug}-${i}`;
    params[0] = candidate;
    const dupeRows = await queryRows(q, checkSql, params);
    if (dupeRows.length === 0) return candidate;
  }

  // Fallback: append random suffix
  const suffix = Math.random().toString(36).substring(2, 8);
  return `${baseSlug}-${suffix}`;
}

async function autoGenerateSlug(
  q: Queryable,
  entity: Entity,
  fields: Record<string, any>,
  isCreate: boolean,
  old: Record<string, any>,
  existingId: any,
): Promise<void> {
  const slugCfg = entity.slug;
  if (!slugCfg || !slugCfg.source) return;

  const slugField = slugCfg.field;
  const sourceField = slugCfg.source;

  // If slug is explicitly provided in the payload, skip auto-generation
  if (fields[slugField] != null && fields[slugField] !== "") return;

  if (isCreate) {
    // On create: generate from source field
    const sourceValue = fields[sourceField];
    if (sourceValue == null || sourceValue === "") return;
    fields[slugField] = await generateUniqueSlug(q, entity, slugify(String(sourceValue)));
  } else if (slugCfg.regenerate_on_update) {
    // On update: regenerate only if source field changed
    const sourceValue = fields[sourceField];
    if (sourceValue == null || sourceValue === "") return;
    if (sourceValue === old[sourceField]) return; // source didn't change
    fields[slugField] = await generateUniqueSlug(q, entity, slugify(String(sourceValue)), existingId);
  }
}

/**
 * For each file-type field whose value is a UUID string, resolve it to
 * the full JSONB metadata object {id, filename, size, mime_type, url}.
 * If the value is already an object, pass through unchanged.
 */
async function resolveFileFields(
  q: Queryable,
  entity: Entity,
  fields: Record<string, any>,
): Promise<void> {
  for (const f of entity.fields) {
    if (f.type !== "file") continue;
    const val = fields[f.name];
    if (val === undefined || val === null) continue;
    // If already an object (full metadata), pass through
    if (typeof val === "object") continue;
    // If it's a UUID string, resolve from _files
    const strVal = String(val);
    if (!UUID_RE.test(strVal)) continue;

    const rows = await queryRows(
      q,
      "SELECT id, filename, size, mime_type FROM _files WHERE id = $1",
      [strVal],
    );
    if (rows.length === 0) {
      throw new AppError("NOT_FOUND", 404, `File ${strVal} not found`);
    }
    const fileRow = rows[0];
    fields[f.name] = {
      id: fileRow.id,
      filename: fileRow.filename,
      size: Number(fileRow.size),
      mime_type: fileRow.mime_type,
    };
  }
}

File: backend/elixir-phoenix/lib/rocket/engine/nested_write.ex
defmodule Rocket.Engine.NestedWrite do
  @moduledoc "Plan-then-execute transactional write pipeline."

  alias Rocket.Store
  alias Rocket.Metadata.{Entity, Registry}
  alias Rocket.Engine.{Writer, AppError, Diff}
  alias Rocket.Instrument.Instrumenter

  @uuid_regex ~r/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
  @int_regex ~r/^\d+$/

  defmodule WritePlan do
    defstruct [:entity, :fields, :id, :user, is_create: true, child_ops: []]
  end

  @doc "Build a WritePlan from request body. Returns {:ok, plan} or {:error, app_error}."
  def plan_write(entity, registry, body, existing_id) do
    {fields, rel_writes, unknown_keys} =
      Writer.separate_fields_and_relations(entity, registry, body)

    if unknown_keys != [] do
      errs =
        Enum.map(unknown_keys, fn key ->
          %{field: key, rule: "unknown", message: "Unknown field or relation: #{key}"}
        end)

      {:error, AppError.validation_failed(errs)}
    else
      is_create = existing_id == nil
      validation_errs = Writer.validate_fields(entity, fields, is_create)

      if validation_errs != [] do
        {:error, AppError.validation_failed(validation_errs)}
      else
        plan = %WritePlan{
          is_create: is_create,
          entity: entity,
          fields: fields,
          id: existing_id,
          child_ops: Map.values(rel_writes)
        }

        {:ok, plan}
      end
    end
  end

  @doc "Execute the write plan inside a single transaction. Returns {:ok, record} or {:error, reason}."
  def execute_write_plan(conn, registry, plan) do
    span = Instrumenter.start_span("engine", "writer", "write.execute_plan")
    span = Instrumenter.set_entity(span, plan.entity.name)

    try do
      result = do_execute_write_plan(conn, registry, plan)

      _span = case result do
        {:ok, _, _} -> Instrumenter.set_status(span, "ok")
        {:error, _} -> Instrumenter.set_status(span, "error")
        _ -> span
      end

      result
    catch
      kind, err ->
        _span = Instrumenter.set_status(span, "error")
        :erlang.raise(kind, err, __STACKTRACE__)
    after
      Instrumenter.end_span(span)
    end
  end

  defp do_execute_write_plan(conn, registry, plan) do
    # We need a raw Postgrex transaction
    execute_in_transaction(conn, fn tx_conn ->
      # Fetch old record for update
      old =
        if !plan.is_create do
          case fetch_record(tx_conn, plan.entity, plan.id) do
            {:ok, row} -> row
            _ -> %{}
          end
        else
          %{}
        end

      # Evaluate rules (stubbed for Phase 0, implemented in Phase 1)
      # Auto-generate slug if configured
      fields = auto_generate_slug(tx_conn, plan.entity, plan.fields, plan.is_create, old, plan.id)
      plan = %{plan | fields: fields}

      with :ok <- evaluate_rules_if_available(registry, plan, old),
           :ok <- evaluate_state_machines_if_available(registry, plan, old),
           :ok <- resolve_file_fields(tx_conn, plan.entity, plan.fields) do
        # Execute parent write
        parent_id =
          if plan.is_create do
            {sql, params} = Writer.build_insert_sql(plan.entity, plan.fields)

            case Store.query_row(tx_conn, sql, params) do
              {:ok, row} -> row[plan.entity.primary_key.field]
              {:error, err} -> throw({:write_error, err})
            end
          else
            {sql, params} = Writer.build_update_sql(plan.entity, plan.id, plan.fields)

            if sql != "" do
              case Store.exec(tx_conn, sql, params) do
                {:ok, _} -> :ok
                {:error, err} -> throw({:write_error, err})
              end
            end

            plan.id
          end

        # Execute child writes
        Enum.each(plan.child_ops, fn child_op ->
          case Diff.execute_child_write(tx_conn, registry, parent_id, child_op) do
            :ok -> :ok
            {:error, err} -> throw({:write_error, err})
          end
        end)

        # Fire sync webhooks (stubbed for Phase 0, implemented in Phase 5)
        fire_sync_webhooks_if_available(tx_conn, registry, plan, old)

        {parent_id, old}
      else
        {:error, %AppError{} = err} -> throw({:app_error, err})
        {:error, err} -> throw({:write_error, err})
      end
    end)
  end

  defp execute_in_transaction(conn, fun) when is_atom(conn) do
    # Ecto.Repo
    conn.transaction(fn ->
      case Ecto.Adapters.SQL.query(conn, "SELECT 1", []) do
        {:ok, _} -> :ok
        _ -> :ok
      end

      fun.(conn)
    end)
    |> handle_tx_result(conn)
  rescue
    e -> {:error, e}
  catch
    {:app_error, err} -> {:error, err}
    {:write_error, {:unique_violation, _}} ->
      {:error, AppError.conflict("A record with this value already exists")}
    {:write_error, err} -> {:error, err}
  end

  defp execute_in_transaction(conn, fun) when is_pid(conn) do
    Store.transaction(conn, fn tx_conn ->
      fun.(tx_conn)
    end)
    |> handle_tx_result(conn)
  rescue
    e -> {:error, e}
  catch
    {:app_error, err} -> {:error, err}
    {:write_error, {:unique_violation, _}} ->
      {:error, AppError.conflict("A record with this value already exists")}
    {:write_error, err} -> {:error, err}
  end

  defp handle_tx_result({:ok, {parent_id, old}}, _conn) do
    {:ok, parent_id, old}
  end

  defp handle_tx_result({:ok, parent_id}, _conn) do
    {:ok, parent_id, %{}}
  end

  defp handle_tx_result({:error, :rollback}, _conn) do
    {:error, AppError.new("INTERNAL_ERROR", 500, "Transaction rolled back")}
  end

  defp handle_tx_result({:error, err}, _conn), do: {:error, err}

  @doc "Execute write plan and return the full record."
  def execute_and_fetch(conn, registry, plan) do
    case execute_write_plan(conn, registry, plan) do
      {:ok, parent_id, old} ->
        # Fetch outside transaction
        case fetch_record(conn, plan.entity, parent_id) do
          {:ok, record} ->
            # Post-commit: trigger workflows for state transitions
            trigger_workflows_if_available(conn, registry, plan, old, record, parent_id)
            # Post-commit: fire async webhooks
            fire_async_webhooks_if_available(conn, registry, plan, old, record)
            {:ok, record}

          err ->
            err
        end

      {:error, _} = err ->
        err
    end
  end

  @doc "Fetch a single record by ID or slug."
  def fetch_record(conn, entity, id_or_slug) do
    columns = Entity.field_names(entity)

    columns =
      if entity.soft_delete && !Entity.has_field?(entity, "deleted_at") do
        columns ++ ["deleted_at"]
      else
        columns
      end

    soft_delete_clause = if entity.soft_delete, do: " AND deleted_at IS NULL", else: ""
    id_str = to_string(id_or_slug)

    # If entity has slug config and param doesn't look like PK type, try slug first
    if entity.slug != nil && is_binary(id_or_slug) && !looks_like_pk?(entity, id_str) do
      slug_sql = "SELECT #{Enum.join(columns, ", ")} FROM #{entity.table} WHERE #{entity.slug.field} = $1#{soft_delete_clause}"

      case Store.query_row(conn, slug_sql, [id_str]) do
        {:ok, row} ->
          {:ok, Store.fix_booleans(row, entity)}

        _ ->
          # slug lookup failed, fall through to PK lookup
          fetch_by_pk(conn, entity, columns, soft_delete_clause, id_or_slug)
      end
    else
      fetch_by_pk(conn, entity, columns, soft_delete_clause, id_or_slug)
    end
  end

  defp fetch_by_pk(conn, entity, columns, soft_delete_clause, id) do
    sql = "SELECT #{Enum.join(columns, ", ")} FROM #{entity.table} WHERE #{entity.primary_key.field} = $1#{soft_delete_clause}"

    case Store.query_row(conn, sql, [id]) do
      {:ok, row} -> {:ok, Store.fix_booleans(row, entity)}
      err -> err
    end
  end

  defp looks_like_pk?(entity, value) do
    case entity.primary_key.type do
      "uuid" -> Regex.match?(@uuid_regex, value)
      t when t in ["int", "integer", "bigint"] -> Regex.match?(@int_regex, value)
      _ -> false
    end
  end

  @doc "Convert text to a URL-friendly slug."
  def slugify(text) when is_binary(text) do
    text
    |> String.downcase()
    |> String.normalize(:nfd)
    |> String.replace(~r/[^\x00-\x7F]/, "")
    |> String.replace(~r/[^a-z0-9]+/, "-")
    |> String.replace(~r/^-+|-+$/, "")
    |> String.replace(~r/-{2,}/, "-")
  end

  def slugify(_), do: ""

  defp generate_unique_slug(conn, entity, base_slug, exclude_id \\ nil) do
    slug_field = entity.slug.field
    soft_delete_clause = if entity.soft_delete, do: " AND deleted_at IS NULL", else: ""
    {exclude_clause, extra_params} =
      if exclude_id != nil do
        {" AND #{entity.primary_key.field} != $2", [exclude_id]}
      else
        {"", []}
      end

    check_sql = "SELECT 1 FROM #{entity.table} WHERE #{slug_field} = $1#{soft_delete_clause}#{exclude_clause} LIMIT 1"

    case Store.query_rows(conn, check_sql, [base_slug | extra_params]) do
      {:ok, []} -> base_slug
      {:ok, _} -> try_slug_suffixes(conn, check_sql, base_slug, extra_params, 2)
      _ -> base_slug
    end
  end

  defp try_slug_suffixes(_conn, _sql, base_slug, _extra_params, n) when n > 100 do
    "#{base_slug}-#{n}"
  end

  defp try_slug_suffixes(conn, check_sql, base_slug, extra_params, n) do
    candidate = "#{base_slug}-#{n}"

    case Store.query_rows(conn, check_sql, [candidate | extra_params]) do
      {:ok, []} -> candidate
      {:ok, _} -> try_slug_suffixes(conn, check_sql, base_slug, extra_params, n + 1)
      _ -> candidate
    end
  end

  defp auto_generate_slug(_conn, %{slug: nil}, fields, _is_create, _old, _existing_id), do: fields
  defp auto_generate_slug(_conn, %{slug: %{source: nil}}, fields, _is_create, _old, _existing_id), do: fields
  defp auto_generate_slug(_conn, %{slug: %{source: ""}}, fields, _is_create, _old, _existing_id), do: fields

  defp auto_generate_slug(conn, entity, fields, is_create, old, existing_id) do
    slug_cfg = entity.slug
    slug_field = slug_cfg.field

    # If slug is explicitly provided, skip auto-generation
    slug_val = Map.get(fields, slug_field)
    if slug_val != nil && slug_val != "" do
      fields
    else
      source_val = Map.get(fields, slug_cfg.source)

      cond do
        source_val == nil || source_val == "" ->
          fields

        is_create ->
          slug = generate_unique_slug(conn, entity, slugify(to_string(source_val)))
          Map.put(fields, slug_field, slug)

        slug_cfg.regenerate_on_update ->
          old_source = Map.get(old, slug_cfg.source, "")
          if to_string(source_val) == to_string(old_source) do
            fields
          else
            slug = generate_unique_slug(conn, entity, slugify(to_string(source_val)), existing_id)
            Map.put(fields, slug_field, slug)
          end

        true ->
          fields
      end
    end
  end

  # Stubs for phases not yet implemented — they succeed silently
  defp evaluate_rules_if_available(registry, plan, old) do
    if Code.ensure_loaded?(Rocket.Engine.Rules) &&
         function_exported?(Rocket.Engine.Rules, :evaluate_rules, 6) do
      action = if plan.is_create, do: "create", else: "update"

      case Rocket.Engine.Rules.evaluate_rules(
             registry,
             plan.entity.name,
             "before_write",
             plan.fields,
             old,
             action
           ) do
        [] -> :ok
        errs -> {:error, AppError.validation_failed(errs)}
      end
    else
      :ok
    end
  end

  defp evaluate_state_machines_if_available(registry, plan, old) do
    if Code.ensure_loaded?(Rocket.Engine.StateMachineEngine) &&
         function_exported?(Rocket.Engine.StateMachineEngine, :evaluate_state_machines, 5) do
      case Rocket.Engine.StateMachineEngine.evaluate_state_machines(
             registry,
             plan.entity.name,
             plan.fields,
             old,
             plan.is_create
           ) do
        [] -> :ok
        errs -> {:error, AppError.validation_failed(errs)}
      end
    else
      :ok
    end
  end

  defp fire_sync_webhooks_if_available(conn, registry, plan, old) do
    if Code.ensure_loaded?(Rocket.Engine.WebhookEngine) &&
         function_exported?(Rocket.Engine.WebhookEngine, :fire_sync_webhooks, 8) do
      action = if plan.is_create, do: "create", else: "update"
      hook = if plan.is_create, do: "before_write", else: "before_write"

      case Rocket.Engine.WebhookEngine.fire_sync_webhooks(
             conn, registry, hook, plan.entity.name, action, plan.fields, old, nil
           ) do
        :ok -> :ok
        {:error, err} -> throw({:write_error, err})
      end
    else
      :ok
    end
  end

  defp trigger_workflows_if_available(conn, registry, plan, old, record, parent_id) do
    if Code.ensure_loaded?(Rocket.Engine.WorkflowEngine) &&
         function_exported?(Rocket.Engine.WorkflowEngine, :trigger_workflows, 7) do
      sms = Registry.get_state_machines_for_entity(registry, plan.entity.name)

      Enum.each(sms, fn sm ->
        old_state = if is_map(old), do: Map.get(old, sm.field, ""), else: ""
        old_state = old_state || ""
        new_state = Map.get(plan.fields, sm.field, "")
        new_state = new_state || ""

        if new_state != "" && old_state != new_state do
          Rocket.Engine.WorkflowEngine.trigger_workflows(
            conn, registry, plan.entity.name, sm.field, new_state, record, parent_id
          )
        end
      end)
    end
  rescue
    e ->
      require Logger
      Logger.error("Workflow trigger error: #{inspect(e)}")
  end

  defp fire_async_webhooks_if_available(conn, registry, plan, old, record) do
    if Code.ensure_loaded?(Rocket.Engine.WebhookEngine) &&
         function_exported?(Rocket.Engine.WebhookEngine, :fire_async_webhooks, 8) do
      action = if plan.is_create, do: "create", else: "update"

      Rocket.Engine.WebhookEngine.fire_async_webhooks(
        conn, registry, "after_write", plan.entity.name, action, record, old, nil
      )
    end
  rescue
    e ->
      require Logger
      Logger.error("Async webhook error: #{inspect(e)}")
  end

  defp resolve_file_fields(conn, entity, fields) do
    file_fields = Enum.filter(entity.fields, &(&1.type == "file"))

    Enum.reduce_while(file_fields, :ok, fn f, :ok ->
      val = Map.get(fields, f.name)

      cond do
        val == nil ->
          {:cont, :ok}

        is_map(val) ->
          {:cont, :ok}

        is_binary(val) && Regex.match?(@uuid_regex, val) ->
          case Store.query_row(
                 conn,
                 "SELECT id, filename, size, mime_type FROM _files WHERE id = $1",
                 [val]
               ) do
            {:ok, row} ->
              Map.put(fields, f.name, %{
                "id" => row["id"],
                "filename" => row["filename"],
                "size" => row["size"],
                "mime_type" => row["mime_type"]
              })

              {:cont, :ok}

            {:error, :not_found} ->
              {:halt, {:error, AppError.not_found("File", val)}}

            {:error, err} ->
              {:halt, {:error, err}}
          end

        true ->
          {:cont, :ok}
      end
    end)
  end
end
